### PR TITLE
Build grids of pre-computed mixing ratios with FastChem

### PR DESCRIPTION
**Pull request by bmorris3**
This PR builds on #5. It:
* adds [narrative documentation for the chemistry module](https://shone--6.org.readthedocs.build/en/6/shone/examples/chemistry.html)
* corrects an error in the assumed units for the VMR calculations
* adds a method for assembling an xarray dataset of FastChem results for a grid of temperatures, pressures, metallicities, and C/O ratios; and saves it to disk
* adds a method that returns a jitted VMR interpolator, so we can move FastChem "offline" during inference
* adds more informative error messages when files in the `~/.shone` dir have unexpected paths

### CODE REVIEW

File: docs/shone/examples/chemistry.rst [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+We compute the matrix of volume mixing ratios for all species with
+`~shone.chemistry.FastchemWrapper.vmr`, and index it for a particular
+species:
+
+.. code-block:: python
+
+    # lookup the column index for O2 in the fastchem VMR matrix:
+    idx = species_table.loc['O2']['index']
+    vmr_O2 = vmr[: idx]
```
Comment (Hoeijmakers) [inline docs/shone/examples/chemistry.rst:122]: This here should be `vmr_O2 = chem.vmr()[: idx]`

File: docs/shone/examples/chemistry.rst [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+.. code-block:: python
+
+    import matplotlib.pyplot as plt
+
+    species = ['H2O1', 'O1Ti1', 'e-']
+    indices = species_table.loc[species]['index']
+    names = species_table.loc[species]['name']
+
+    ax = plt.gca()
+    ax.loglog(vmr[:, indices], pressure, label=names)
```
Comment (Hoeijmakers) [inline docs/shone/examples/chemistry.rst:135]: And: `ax.loglog(chem.vmr()[:, indices], pressure, label=names)`

File: shone/chemistry/fastchem.py [hunk 3]
```diff
@@ -201,7 +217,180 @@ def mmr_mmw(self):
...
+        mmr_mmw = vmr * weights_amu[None, :]
 
         return mmr_mmw
+
+
+def round_in_log(x):
+    log_floor = np.floor(np.log10(x))
+    pre_exponent = np.round(x * 10 ** -log_floor, 1)
+    return pre_exponent * 10 ** log_floor
+
```
Comment (Hoeijmakers) [inline shone/chemistry/fastchem.py:230]: Should this be jittable and does it need to be jnp?

File: docs/shone/examples/chemistry.rst [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+.. code-block:: python
+
+    from shone.chemistry import build_fastchem_grid
+
+    chem_grid = build_fastchem_grid()
+
+The grid is saved to your `~/.shone` directory and can be interpolated during sampling to
+use *approximate* FastChem mixing ratios.
+
```
Comment (Hoeijmakers) [inline docs/shone/examples/chemistry.rst:215]: State here what the defaults are? Temperature between 300 to 6000 in 22 steps, pressure logarithmically from 1e-8 to 10 bars in 20 steps, metallicity from -1 to 2.5 in 8 steps; log(c/o) from -1 to 2. 

Is that, by the way, a sensible range? When we say solar C/O is 0.52; that's not log(C/O), right?

File: docs/shone/examples/chemistry.rst [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+Interpolate from the grid
++++++++++++++++++++++++++
+
+Now let's print a table of volume mixing ratios for the first five species:
```
Comment (Hoeijmakers) [inline docs/shone/examples/chemistry.rst:219]: It's not clear here in the docs how the chem_grid object of the previous cell is used here.
The jitted chemistry interpolator here below on line 227 is automatically using that fastchem grid?

**Review by Hoeijmakers — CHANGES_REQUESTED** [review]
Looking good! I had a few small issues that I commented in the change-file.

Also, when building the docs locally I get warnings about automatically generated docfiles not being linked. Is that a problem?

image

File: shone/chemistry/fastchem.py (hunk 3 above)
Comment (bmorris3) [reply shone/chemistry/fastchem.py:230]: It doesn't need to be, it's just a convenience function for getting round and roughly log-spaced numbers from a true log-spaced array.

**Comment by bmorris3** [comment]
RE @Hoeijmakers:
> Also, when building the docs locally I get warnings about automatically generated docfiles not being linked. Is that a problem?

These aren't a problem. If you delete your `docs/api` dir (which is ok to do!) and run it again, these should go away.

File: docs/shone/examples/chemistry.rst (hunk 5 above)
Comment (bmorris3) [reply docs/shone/examples/chemistry.rst:219]: The above output was not used directly here – that step builds the grid and saves it to disk, and the code block that you highlighted here loads it from disk. I added a comment to make this clearer.

File: docs/shone/examples/chemistry.rst (hunk 2 above)
Comment (bmorris3) [reply docs/shone/examples/chemistry.rst:135]: Thanks!

**Pull request merged**